pkgs/build-support/pkg-config-wrapper/setup-hook.sh

```diff
@@ -25,5 +25,11 @@ addEnvHooks "$targetOffset" pkgConfigWrapper_addPkgConfigPath
 
 export PKG_CONFIG${role_post}=@targetPrefix@@baseBinName@
 
+# The Rust pkg-config crate does not support prefixed pkg-config executables[1],
+# but it does support checking these idiosyncratic PKG_CONFIG_${TRIPLE}
+# environment variables.
+# [1]: https://github.com/rust-lang/pkg-config-rs/issues/53
+export PKG_CONFIG_@rustTarget@=@targetPrefix@@baseBinName@
```

Comment:
I can't really approve of this rust-specific thing being added in all derivations that use pkg-config, even non-rust ones. I also don't disapprove, but for a quick merge you could move this change out of this PR and let me review only the cargo stuff (with that `PKG_CONFIG_` variable being set just like in rustc).

Reply:
What if it was in the rustc setuphook? Than as soon as you add rustc to a build, you would get it.

Reply:
I think the pkg-config wrapper is the right place for this.  It's the only place that makes sense that is shared between the rustc build, and all the builds that use Cargo.

Technically, it doesn't have anything to do with pkg-config (since it comes from the pkg-config crate), but it doesn't have anything to do with Cargo either.  It's also extremely similar to PKG_CONFIG_FOR_BUILD, defined on the line above, so it makes sense to set both very similar environment variables in one place.

Reply:
I agree it doesn't go here.

We can have a separate setup hook that Rust. `buildRustCrate`, `buildRustPackage`, etc. all use.

Reply:
https://github.com/rust-lang/pkg-config-rs/issues/139 Also, I opened an issue in the relevant package `build.rs` file use.

Reply:
I don't want more setup hooks that have to be manually added to a build — it makes it too complicated for package authors to do the right thing, especially when it's not obvious they need to do it, like in this case where only cross-compilation is affected.  So what package could bring this in, if not for pkg-config?  cargo?  This env var is even less related to cargo than it is to pkg-config.

Reply:
And what's the justification for letting `PKG_CONFIG_FOR_BUILD` be set here and not the pkg-config-rs environment variables?

`PKG_CONFIG_FOR_BUILD` doesn't come from the pkg-config package either, at least as far as I can tell.  It's an autoconf/Meson/etc thing, so if we were to treat it as you're suggesting we do the pkg-config-rs variables, we'd have to move it into some other setup hook too.

Reply:
See what I left in the upstream stuff, but I really rather have it be a separate setup hook. There are many such things like Cargo/build.rs I don't think it's write to support them all in the tool wrappers. M tools and N consumption patterns I think should be handled separately so one "pays for just what one uses" and complexity is kept under control.

Reply:
I think this is premature.  There are two conventions, and for each one it's setting a single environment variable.  "Paying for what you use" doesn't make sense when the cost is negligible.

Reply:
It's a small amount of code, but changing it is mass rebuild. So I think about the "mass rebuild" costs of these sorts of things too.

Reply:
Also think how if we do two setup hooks it's easy to combine later, but if we do one setup hook it's hard to split later.

Reply:
That would be an extremely rare change to make, especially for the limited set of architectures we actually build for.  We do full rebuilds every couple of weeks anyway, so I don't think it's at all worth optimising for avoidance of mass-rebuilds in infrastructural stuff like this.  Indeed, avoiding mass rebuilds is how we end up with unnecessary conditionals making maintenance more complicated forever, because it avoided mass rebuilds once.  And Rust is so widely used that any change to it is going to cause huge rebuilds anyway, regardless of whether it affects pkg-config.

Reply:
And having a separate setup hook just for this (`cargoPkgConfigSetupHook`?) would mean broken cross packages forever, because since it would only be required for cross, nobody would ever remember to apply it.